sh/helpers_test.go

```diff
@@ -0,0 +1,140 @@
+package sh_test
+
+import (
+	"bytes"
+	"fmt"
+	"io/ioutil"
+	"math/rand"
+	"os"
+	"path/filepath"
+	"testing"
+
+	"github.com/magefile/mage/sh"
+)
+
+func makeTestFile(dir string, name string) (string, error) {
+	fname := filepath.Join(dir, name)
+	f, err := os.Create(fname)
+	if err != nil {
+		return fname, fmt.Errorf("can't create test file %s: %v", fname, err)
+	}
+	n := rand.Intn(255) + 1
+	for i := 0; i < n; i++ {
+		f.Write([]byte("All work and no play makes Jack a dull boy.\n"))
+	}
+	err = f.Close()
+	if err != nil {
+		return fname, fmt.Errorf("error closing test file %s: %v", fname, err)
+	}
+	return fname, nil
+}
+
+// compareFiles checks that two files are identical for testing purposes. That means they have the same length,
+// the same contents, and the same permissions. It does NOT mean they have the same timestamp, as that is expected
+// to change in normal Mage sh.Copy operation.
+func compareFiles(file1 string, file2 string) error {
+	s1, err := os.Stat(file1)
+	if err != nil {
+		return fmt.Errorf("can't stat %s: %v", file1, err)
+	}
+	s2, err := os.Stat(file2)
+	if err != nil {
+		return fmt.Errorf("can't stat %s: %v", file2, err)
+	}
+	if s1.Size() != s2.Size() {
+		return fmt.Errorf("files %s and %s have different sizes: %d vs %d", file1, file2, s1.Size(), s2.Size())
+	}
+	if s1.Mode() != s2.Mode() {
+		return fmt.Errorf("files %s and %s have different permissions: %#4o vs %#4o", file1, file2, s1.Mode(), s2.Mode())
+	}
+	f1bytes, err := ioutil.ReadFile(file1)
+	if err != nil {
+		return fmt.Errorf("can't read %s: %v", file1, err)
+	}
+	f2bytes, err := ioutil.ReadFile(file2)
+	if err != nil {
+		return fmt.Errorf("can't read %s: %v", file2, err)
+	}
+	cmp := bytes.Compare(f1bytes, f2bytes)
+	if cmp != 0 {
+		return fmt.Errorf("files %s and %s have different contents", file1, file2)
+	}
+	return nil
+}
+
+func TestHelpers(t *testing.T) {
+
+	mytmpdir, err := ioutil.TempDir(os.TempDir(), "mage")
+	if err != nil {
+		t.Errorf("can't create test directory: %v", err)
+		return
+	}
+	defer func() {
+		derr := os.RemoveAll(mytmpdir)
+		if derr != nil {
+			fmt.Printf("error cleaning up after TestHelpers: %v", derr)
+		}
+	}()
+	srcname, err := makeTestFile(mytmpdir, "test1.txt")
+	if err != nil {
+		t.Error(err)
+	}
+	destname := filepath.Join(mytmpdir, "test2.txt")
+
+	t.Run("sh/copy", func(t *testing.T) {
+		cerr := sh.Copy(destname, srcname)
```

Comment:
so, I don't think it's super useful to name errors something other than `err` unless it's actually necessary to distinguish between them (and it's not here).  It kind of makes the code harder to read, because when I see `if cerr != nil {` I have to think about what type `cerr` is.  Whereas `if err != nil {` might as well be one complete token, because it's so common in go, everyone knows what it means immediately.

Reply:
I do it to avoid warnings about variable shadowing when I lint the code.

Reply:
IMO variable shadowing isn't a bad thing.  Yes it can sometimes cause errors, but that's pretty rare in my experience, and it makes the code a lot easier to read precisely because you don't need to give every error a different name.  Then I don't have to think about if this is cerr or derr... it's just always err.  ¯\\_(ツ)_/¯ .  Your mileage may vary.